File: app/src/main/java/example/homework/Adapter.java
package example.homework;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

public abstract class Adapter extends RecyclerView.Adapter<Adapter.ViewHolder> {

    private LayoutInflater inflater;
    private List<Item> items;

    public Adapter(Context context, List<Item> items) {
        this.inflater = LayoutInflater.from(context);
        this.items = items;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.name.setText(items.get(position).getName());
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        private final TextView name;
        private View.OnClickListener clickListener;

        ViewHolder(View itemView) {
            super(itemView);
            name = itemView.findViewById(R.id.tv_name);
            clickListener = new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    onItemClicked(getAdapterPosition());
                }
            };
            itemView.setOnClickListener(clickListener);
        }
    }

    public abstract void onItemClicked(int position);
}

File: app/src/main/java/example/homework/Adapter.kt
package example.homework

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

internal abstract class Adapter(context: Context, private val items: List<Item>) : RecyclerView.Adapter<Adapter.ViewHolder>() {

    private val inflater: LayoutInflater = LayoutInflater.from(context)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = inflater.inflate(R.layout.item, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.name.text = items[position].name
    }

    override fun getItemCount(): Int {
        return items.size
    }

    internal inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        internal val name: TextView = itemView.findViewById(R.id.tv_name)
        private val clickListener: View.OnClickListener = View.OnClickListener { onItemClicked(adapterPosition) }

        init {
            itemView.setOnClickListener(clickListener)
        }
    }

    abstract fun onItemClicked(position: Int)
}
